Layoutlmv2 tesseractconfig

kelvinAI opened the pull request:
# What does this PR do?

Giving user option to set config parameter used by Tesseract when performing feature extraction. Eg. to change psm levels while performing transcription by passing in '--psm 10' to config parameter while invoking image_to_data

It is shown that changing the psm values greatly influences the end result of LayoutLMV2/XLM/V3, and the specific psm value is different depending on the document formatting. Refer : [PSM](https://github.com/tesseract-ocr/tesseract/issues/434)

```python
pytesseract.image_to_data(image, lang=lang, output_type="dict", config="--psm 10") 
``` 

Users can now set the tesseract config parameter during Processor initialization, like so:
```python
# LayoutLMV2
processor = LayoutLMv2Processor.from_pretrained("microsoft/layoutlmv2-base-uncased", ocr_lang="eng", tesseract_config="--psm 5")

# LayoutLMV3
processor = AutoProcessor.from_pretrained("microsoft/layoutlmv3-base", ocr_lang="eng", tesseract_config="--psm 5")
```


## Before submitting
- [❌] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [✔️] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [❌] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [✔️] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [❌] Did you write any new necessary tests?

kelvinAI committed:
Added option for users to modify config parameter used by pytesseract during feature extraction

- Added optional 'tess_config' kwarg when setting up LayoutLMV2 processor that is used by pytesseract during feature extraction
- Eg. Can be used to modify psm values by setting tess_config to '--psm 7'
- Different psm values significantly influences the output of layoutlmv2

kelvinAI committed:
Update src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

kelvinAI committed:
Update src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

kelvinAI committed:
Updated variable names to be more explicit

kelvinAI committed:
Fixed styles

kelvinAI commented:
@NielsRogge kindly review. Thanks!

NielsRogge reviewed (approved):
LGTM! Could you add it to LayoutLMv3's feature extractor as well?

kelvinAI commented:
> LGTM! Could you add it to LayoutLMv3's feature extractor as well?

Sounds good! I'll ping you when it's done. Do I need to do anything to merge this to huggingface:main? You'll trigger the merge is that right?

NielsRogge commented:
>You'll trigger the merge is that right?

Yes, indeed.

kelvinAI committed:
Added option for users to modify config parameter when calling pytesseract during feature extraction

- Added option to set "tesseract_config" parameter during LayoutLMV3 processor initialization
- Can be used to modify PSM values, eg. by setting tesseract_config="--psm 6"

kelvinAI committed:
Merge branch 'layoutlmv3_tesseractconfig' into layoutlmv2_tesseractconfig

kelvinAI commented:
> Hi @kelvinAI, could you add it to LayoutLMv3 as well in this PR?
> 
> Thanks!

Done!

NielsRogge commented on src/transformers/models/layoutlmv2/feature_extraction_layoutlmv2.py:
```diff
@@ -103,6 +103,9 @@ class LayoutLMv2FeatureExtractor(FeatureExtractionMixin, ImageFeatureExtractionM
         ocr_lang (`Optional[str]`, *optional*):
             The language, specified by its ISO code, to be used by the Tesseract OCR engine. By default, English is
             used.
+        tesseract_config (`Optional[str]`, *optional*):
```
```suggestion
        tesseract_config (`str`, *optional*):
```
You can actually also remove "Optional" from the `ocr_lang` argument above (can't comment there)

NielsRogge commented on src/transformers/models/layoutlmv3/feature_extraction_layoutlmv3.py:
```diff
@@ -109,6 +108,9 @@ class LayoutLMv3FeatureExtractor(FeatureExtractionMixin, ImageFeatureExtractionM
         ocr_lang (`Optional[str]`, *optional*):
             The language, specified by its ISO code, to be used by the Tesseract OCR engine. By default, English is
             used.
+        tesseract_config (`Optional[str]`, *optional*):
```
```suggestion
        tesseract_config (`str`, *optional*):
```
Same here

NielsRogge commented:
Hi @kelvinAI, could you apply the suggestions such that I can merge your PR?

Thanks!

kelvinAI committed:
Removed  from function signature

kelvinAI commented:
@NielsRogge done! :)

sgugger reviewed (approved):
Thanks for adding this!
